File: clojure/allergies/src/allergies.clj
(ns allergies)

(def allergen [:eggs
               :peanuts
               :shellfish
               :strawberries
               :tomatoes
               :chocolate
               :pollen
               :cats])

(defn allergies [score]
  (reduce-kv (fn [accu bit item]
               (if (bit-test score bit) (conj accu item) accu))
             []
             allergen))

(defn allergic-to? [score item]
  (some #{item} (allergies score)))

File: haskell/allergies/src/Allergies.hs
module Allergies (Allergen(..), allergies, isAllergicTo) where

import Data.Bits (testBit)

data Allergen = Eggs
              | Peanuts
              | Shellfish
              | Strawberries
              | Tomatoes
              | Chocolate
              | Pollen
              | Cats
              deriving (Eq, Enum)

allergies :: Int -> [Allergen]
allergies score = filter (`isAllergicTo` score) $ enumFromTo Eggs Cats

isAllergicTo :: Allergen -> Int -> Bool
isAllergicTo allergen = (`testBit` fromEnum allergen)
